### PR TITLE
New SDK IA (Node) - Usage Examples

### PR DESCRIPTION
**Pull request by nlarew**
## Pull Request Info

All implemented/portable Usage Examples for the Node SDK

Still TODO:
- Fundamentals
- Advanced Guides
- Data Types (non-blocking)

### Issue JIRA link:

(DOCSP-13773): Update TOC & Pages to New IA (Node)

### Docs staging link (requires sign-in on MongoDB Corp SSO):

- [ 👍 ] [Open & Close a Local Realm](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/open-and-close-a-local-realm)
- [👍 ] [Read & Write Data](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/read-and-write-data)
- [ 👍 ] [React to Changes](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/react-to-changes)
- [ 👍 ] [Authenticate Users](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/authenticate-users)
- [👍  ] [Manage Email/Password Users](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/manage-email-password-users)
- [ 👍 ] [Create & Manage User API Keys](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/manage-user-api-keys)
- [ 👍 ] [Access a Linked MongoDB Cluster](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/access-a-linked-mongodb-cluster)
- [ 👍 ] [Call a Function](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/new-ia/sdk/node/examples/call-a-function)

### CODE REVIEW

File: source/sdk/node/examples/open-and-close-a-local-realm.txt [hunk 1]
```diff
@@ -4,22 +4,54 @@
...
+<Realm.html#.open>` or :js-sdk:`new Realm() <Realm.html>`. The following example
+creates a ``Configuration`` object with a path property that defines the local
+file path to store data. 
+
+.. code-block:: javascript
+   
+   const realm = await Realm.open({
+      path: "myRealm",
+      schema: [
+         schemas.TaskSchema,
```
Comment (mongomoe) [inline source/sdk/node/examples/open-and-close-a-local-realm.txt:31]: instead of `schemas.TaskSchema` (and the others) can this just be `TaskSchema`? Also can we put a comment that tells users to define this variable?

File: source/sdk/node/examples/read-and-write-data.txt [hunk 2]
```diff
@@ -22,35 +22,112 @@ Read Operations
 Find a Specific Object by Primary Key
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. _node-query-all-objects-of-a-given-type:
+If you know the :ref:`primary key <node-primary-key>` for a given object, you
+can look it up directly with :js-sdk:`Realm.objectForPrimaryKey()
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:26]:
```suggestion
can look it up directly with :js-sdk:`realm.objectForPrimaryKey()
```
required change: ``objectForPrimaryKey() is a method on the realm instance not on the Realm import.

File: source/sdk/node/examples/read-and-write-data.txt [hunk 3]
```diff
@@ -22,35 +22,112 @@ Read Operations
...
+<Realm.html#objectForPrimaryKey>`.
+
+.. tabs-realm-languages::
+   
+   .. tab::
+      :tabid: javascript
+      
+      .. code-block:: javascript
+         
+         const task = Realm.objectForPrimaryKey("Task", "600115c589ff8d461af84727");
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:36]:
```suggestion
         const task = realm.objectForPrimaryKey("Task", "600115c589ff8d461af84727");
```

required change: ``objectForPrimaryKey() is a method on the realm instance not on the Realm import.  We should probably add a comment that explains this, saying that you must open a realm first, and then use that variable.

File: source/sdk/node/examples/read-and-write-data.txt (hunk 2 above)
Comment (nlarew) [reply source/sdk/node/examples/read-and-write-data.txt:26]: You're right that `objectForPrimaryKey()` is an instance method but it's on the `Realm` class so the capitalized notation is correct

File: source/sdk/node/examples/read-and-write-data.txt [hunk 4]
```diff
@@ -22,35 +22,112 @@ Read Operations
...
+
+.. _node-query-an-object-type:
+.. _node-get-all-objects:
+
+Query an Object Type
+~~~~~~~~~~~~~~~~~~~~
+
+Query operations return a :js-sdk:`Realm.Results <Realm.Results.html>`
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:44]: Nitpick/not required for lgtm: This sentence is a little hard to follow. Although it's not a long one, its easy to get distracted by the term 'Realm.Results' upfront, here's a suggestion that conveys the same concepts but with 'Realm.Results' at the end of the sentence:

`Query operations return a collection that contains all objects that match the query via the Realm.Results type.`

File: source/sdk/node/examples/read-and-write-data.txt [hunk 5]
```diff
@@ -22,35 +22,112 @@ Read Operations
...
+   - Find tasks assigned to specific teammates Ali or Jamie by seeing if the ``assignee`` property is in a list of names.
 
-Filter Queries Based on Object Properties
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   .. tabs-realm-languages::
+
+      .. tab::
+         :tabid: javascript
+
+          .. literalinclude:: /examples/Query/Comparison.js
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:88]: Nitpick/not required for lgtm: Looks like `/examples/Query/Comparison.js` isn't apart of this PR, but can we change this code to have string interpolation rather than concatenation to be consistent with the rest of the node docs.

File: source/sdk/node/examples/read-and-write-data.txt [hunk 6]
```diff
@@ -22,35 +22,112 @@ Read Operations
...
+.. _node-sort-results:
 
 Sort Query Results
 ~~~~~~~~~~~~~~~~~~
 
-.. _node-query-a-relationship:
+A **sort** operation allows you to configure the order in which
+{+client-database+} returns queried objects. You can sort based on one or more
+properties of the objects in the results collection. {+client-database+} only
+guarantees a consistent order of results when the results are sorted.
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:107]:
```suggestion
guarantees a consistent order of results when you call `sort()`
```

Nitpick/not required for lgtm: "To avoid passive voice and be a bit more clear as to who is sorting the results (the developer or mongodb)

File: source/sdk/node/examples/read-and-write-data.txt [hunk 7]
```diff
@@ -62,27 +139,170 @@ Write Operations
...
+         :language: javascript
+         :emphasize-lines: 4, 11
+
+.. _node-update-a-collection:
+.. _node-bulk-update:
+
+Bulk Update a Collection
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+To apply an update to a collection of objects, iterate through the collection
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:208]: Nitpick/not required for LGTM:

Can this sentence be reworded to be less runon? It's a bit hard to follow currently since it lists a few steps in 1 sentence.

File: source/sdk/node/examples/read-and-write-data.txt [hunk 8]
```diff
@@ -62,27 +139,170 @@ Write Operations
...
+   .. tab::
+      :tabid: javascript
+      
+      .. literalinclude:: /examples/CRUD/Delete.js
+         :language: javascript
+         :emphasize-lines: 3
+
+.. important:: Do not use objects after delete
+   
+   {+client-database+} throws an error if you try to use an object after it has
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:246]: Nitpick/not required for LGTM: Passive voice here. Instead of saying "it has been deleted" (which may confuse the reader, as it is not clear who is doing the deleting), can you rephrase it to something like:

"... if you try to use an object that you have deleted"

File: source/sdk/node/examples/read-and-write-data.txt (hunk 7 above)
Comment (nlarew) [reply source/sdk/node/examples/read-and-write-data.txt:208]: Is there a particular bit that threw you off? Not entirely sure what to change, though I will add a link to [forEach()](Web/JavaScript/Reference/Global_Objects/Array/forEach) as an example of a collection iterator

**Review by mongomoe — COMMENTED** [review]
"Read & Write Data" has a couple of required changes needed, but all of the other comments are small suggestions. This page looks really good so far and is close to LGTM.

File: source/sdk/node/examples/read-and-write-data.txt (hunk 8 above)
Comment (nlarew) [reply source/sdk/node/examples/read-and-write-data.txt:246]: To me this is one of the rare cases where some passive voice is okay in that it doesn't matter who deleted the object, just that the object was deleted. The active voice here is "Realm throws an error" & "you use an object". Thoughts?

(I may tweak this a bit regardless)

File: source/includes/mongodb-data-access-example-plant-documents.rst [hunk 9]
```diff
@@ -1,6 +1,6 @@
-The following examples operate on a MongoDB collection that describes
-inventory in a chain of plant stores. Consider the following collection
-of documents describing various plants for sale in store:
+The examples on this page use a MongoDB collection that describes inventory in a
+chain of plant stores. The following collection of documents describe various
+plants for sale in the chain's stores:
```
Comment (ccho-mongodb) [inline source/includes/mongodb-data-access-example-plant-documents.rst:3]: Seems like the content of the collection is described twice, so perhaps you can avoid that by combining them:

```suggestion
The examples on this page use the following MongoDB collection that describes various plants for sale in a chain of plant stores:```

File: source/sdk/node/examples/authenticate-users.txt [hunk 10]
```diff
@@ -14,46 +14,304 @@ Authenticate Users
 
 
 .. _node-anonymous-login:
+.. _node-login-anonymous:
 
 Log In an Anonymous User
 ------------------------
 
+The :doc:`Anonymous </authentication/anonymous>` provider allows users to log in
+to your application with ephemeral accounts that have no associated information.
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/authenticate-users.txt:23]: No change necessary, but I wonder if "temporary", instead of "ephemeral" would be an easier adjective to comprehend. Is "ephemeral account" a normal term used throughout Realm docs?

File: source/sdk/node/examples/call-a-function.txt [hunk 11]
```diff
@@ -16,3 +16,55 @@ Call a Function
 
 Call a Function by Name
 -----------------------
+
+To call a function, you can either pass its name and arguments to
+``User.callFunction()`` or call the function as if it were a method on the
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/call-a-function.txt:21]:
```suggestion
``User.callFunction()`` or call the function as if it was a method on the
```

File: source/sdk/node/examples/call-a-function.txt [hunk 12]
```diff
@@ -16,3 +16,55 @@ Call a Function
...
+:js-sdk:`User.functions <Realm.User.html#functions>` property.
+
+.. note:: Link a MongoDB Atlas Data Source
+   
+   This example requires a MongoDB Realm Application with a :doc:`linked
+   Atlas data source </mongodb/link-a-data-source>`. Replace the
+   "<appId>" in the code block with your app id, available in the
+   lefthand corner of the MongoDB Realm UI.
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/call-a-function.txt:29]: Nitpick: Should we use "lefthand" or "left-hand"? Looks like most instances in Realm docs use "left-hand"

File: source/sdk/node/examples/call-a-function.txt [hunk 13]
```diff
@@ -16,3 +16,55 @@ Call a Function
...
+
+.. code-block:: javascript
+  :emphasize-lines: 21, 22
+
+   const Realm = require("realm");
+
+   const appId = "<appId>"; // Set Realm app ID here.
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/call-a-function.txt:36]:
```suggestion
   const appId = "<appId>"; // Set your Realm app ID here.
```

File: source/sdk/node/examples/call-a-function.txt [hunk 14]
```diff
@@ -16,3 +16,55 @@ Call a Function
...
+
+To call a function, you can either pass its name and arguments to
+``User.callFunction()`` or call the function as if it were a method on the
+:js-sdk:`User.functions <Realm.User.html#functions>` property.
+
+.. note:: Link a MongoDB Atlas Data Source
+   
+   This example requires a MongoDB Realm Application with a :doc:`linked
+   Atlas data source </mongodb/link-a-data-source>`. Replace the
+   "<appId>" in the code block with your app id, available in the
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/call-a-function.txt:28]: Just wanted to suggest consistency with the code comment on line 36
```suggestion
   "<appId>" in the code block with your app ID, available in the
```

File: source/sdk/node/examples/call-a-function.txt [hunk 15]
```diff
@@ -16,3 +16,55 @@ Call a Function
...
+       user.logOut();
+     }
+   }
+   run().catch(console.dir);
+ 
+
+When you run the code sample, your output should resemble the following:
+
+.. code-block:: none
+
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/call-a-function.txt:68]:
```suggestion
    :copyable: false
```

File: source/sdk/node/examples/access-a-linked-mongodb-cluster.txt [hunk 16]
```diff
@@ -0,0 +1,665 @@
...
+   .. tab::
+      :tabid: javascript
+      
+      .. literalinclude:: /examples/generated/code/start/mongodb.codeblock.plants-collection-handle.js
+         :language: javascript
+   
+   .. tab::
+      :tabid: typescript
+      
+      .. literalinclude:: /examples/generated/code/start/mongodb.codeblock.plants-collection-handle.ts
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/access-a-linked-mongodb-cluster.txt:44]: Not part of this PR, but does the Typescript formatter suggest the additional linebreaks not present in the Javascript version? It could be less jarring to keep the formatting consistent when possible.

File: source/sdk/node/examples/access-a-linked-mongodb-cluster.txt [hunk 17]
```diff
@@ -0,0 +1,665 @@
...
+
+Find a Single Document
+~~~~~~~~~~~~~~~~~~~~~~
+
+To find a single document, pass a query filter that matches the document to
+:js-sdk:`collection.findOne() <Realm.MongoDBCollection.html#findOne>`. If you do
+not pass a query filter, ``findOne()`` matches the first document it finds in
+the collection.
+
+The following snippet finds the document that describes venus flytraps in the
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/access-a-linked-mongodb-cluster.txt:61]: Just to clarify that it's a type of plant for anyone that might not have heard of them.

```suggestion
The following snippet finds the document that describes venus flytrap plants in the
```

File: source/sdk/node/examples/access-a-linked-mongodb-cluster.txt [hunk 18]
```diff
@@ -0,0 +1,665 @@
...
+
+   .. tab::
+      :tabid: typescript
+      
+      .. literalinclude:: /examples/generated/code/start/mongodb.codeblock.find-multiple-documents.ts
+         :language: typescript
+
+Running this snippet produces output resembling the following:
+
+.. tabs-realm-languages::
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/access-a-linked-mongodb-cluster.txt:126]: Nitpick: while not related to this PR, I wonder if the output code blocks would be easier to read without the copy widget (`:copyable: false`)

File: source/sdk/node/examples/access-a-linked-mongodb-cluster.txt [hunk 19]
```diff
@@ -0,0 +1,665 @@
...
+         :language: typescript
+
+.. _node-update-a-single-document:
+.. _node-mongodb-updateOne:
+
+Update a Single Document
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+To update a single document, pass a query filter that matches the document and
+an update description to :js-sdk:`collection.updateOne()
```
Comment (ccho-mongodb) [inline source/sdk/node/examples/access-a-linked-mongodb-cluster.txt:284]:
```suggestion
an update document to :js-sdk:`collection.updateOne()
```

File: source/sdk/node/examples/call-a-function.txt (hunk 12 above)
Comment (nlarew) [reply source/sdk/node/examples/call-a-function.txt:29]: We actually have a source constant for this from another time you caught this!

i.e. `{+leftnav+}` which evaluates to "left navigation menu"

File: source/sdk/node/examples/react-to-changes.txt [hunk 20]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+anything changes. As the app becomes more sophisticated and
+performance-sensitive, the app developers shift to more
+granular notifications.
+
+.. example::
+
+  Suppose you are writing a real-time collaborative app. To
+  give the sense that your app is buzzing with collaborative
+  activity, you want to have an indicator that lights up when
+  any change is made. In that case, a realm notification
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:38]: Nitpick/not required for LGTM:
"Any change is made" is very vague and passive voice. What is a 'change', who is making this change?

Suggestion:
"To give the sense that your app is buzzing with collaborative activity, you want to have an indicator that lights up when a user performs a write transaction"

File: source/sdk/node/examples/react-to-changes.txt [hunk 21]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+          :language: javascript
+          :emphasize-lines: 6
+
+.. _node-register-a-collection-change-listener:
 
 Register a Collection Change Listener
 -------------------------------------
 
-.. _register-an-object-change-listener:
+You can register a notification handler on a specific
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:56]: Question/not required for LGTM: 

The content under 'Register a Collection Change Listener' seems  like a lot of explanation for a usage example, is there anyway we can reduce this?

File: source/sdk/node/examples/react-to-changes.txt [hunk 22]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+
+Unlike Realm notifications, collection notifications contain
+detailed information about the change. This enables
+sophisticated and selective reactions to changes. Collection
+notifications provide all the information needed to manage a
+list or other view that represents the collection in the UI.
+
+.. example::
+
+  The following code shows how to observe a collection for
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:86]: Nitpick/not required for LGTM: The code below this doesn't update any UI. Although it may be inferred, this may trip up users. Let's change this sentence to: 

"The following code shows how to observe a collection for changes."

File: source/sdk/node/examples/react-to-changes.txt [hunk 23]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+.. _node-register-an-object-change-listener:
 
 Register an Object Change Listener
 ----------------------------------
 
-.. _unregister-a-change-listener:
+You can register a notification handler on a specific object
+within a Realm. Realm Database notifies your handler:
+
+- When the object is deleted.
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:106]: Nitpick/not required for LGTM:

Passive voice here: "When the object is deleted". Who is deleting the object?

File: source/sdk/node/examples/react-to-changes.txt [hunk 24]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+          :language: javascript
+          :emphasize-lines: 13
+
+.. _node-unregister-a-change-listener:
 
 Unregister a Change Listener
 ----------------------------
+
+To remove all registered realm, object, and collection listeners associated with
+a realm, call :js-sdk:`Realm.removeAllListeners()
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:133]:
```suggestion
a realm, call :js-sdk:`realm.removeAllListeners()
```
It should be the realm instance, not the "Realm" import.

File: source/sdk/node/examples/react-to-changes.txt [hunk 25]
```diff
@@ -12,22 +12,151 @@ React to Changes
...
+<Realm.html#removeAllListeners>`. 
+
+To remove a specific listener, call :js-sdk:`Object.removeListener()
```
Comment (mongomoe) [inline source/sdk/node/examples/react-to-changes.txt:136]:
```suggestion
To remove a specific listener, call :js-sdk:`Realm.Object.removeListener()
```

Required change for LGTM: "Object" is vague, this should either be `"<RealmObject>".removeListener()` or `Realm.Object.removeListener()`. In other instances in the docs we use `"<Thing that needs to be replaced by your variable name>"`.

**Review by mongomoe — COMMENTED** [review]
A couple required changes needed for "React to Changes", and a few non required suggestions and questions, otherwise this is close to LGTM

File: source/examples/Query/Comparison.js [hunk 26]
```diff
@@ -1,10 +1,9 @@
-console.log("Number of high priority tasks: "
-  + tasks.filtered("priority > 5").length);
+const highPriorityTasks = tasks.filtered("priority > 5");
+const unassignedTasks = tasks.filtered("priority > 5");
+const lowProgressTasks = tasks.filtered("1 <= progressMinutes && progressMinutes < 10");
 
 console.log(
```
Comment (mongomoe) [inline source/examples/Query/Comparison.js:5]: Nice!

File: source/sdk/node/examples/read-and-write-data.txt (hunk 7 above)
Comment (mongomoe) [reply source/sdk/node/examples/read-and-write-data.txt:208]: Weird, I see the mdn link change in the PR, but not in the staging link. Did a command+r too

File: source/sdk/node/examples/read-and-write-data.txt [hunk 27]
```diff
@@ -62,27 +139,172 @@ Write Operations
...
+         :emphasize-lines: 4, 11
+
+.. _node-update-a-collection:
+.. _node-bulk-update:
+
+Bulk Update a Collection
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+To apply an update to a collection of objects, iterate through the collection
+(e.g. with :mdn:`forEach()
```
Comment (mongomoe) [inline source/sdk/node/examples/read-and-write-data.txt:209]:
```suggestion
(e.g., with :mdn:`forEach()
```

"e.g.," is the correct punctuation here. https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/

File: source/sdk/node/examples/read-and-write-data.txt (hunk 7 above)
Comment (mongomoe) [reply source/sdk/node/examples/read-and-write-data.txt:208]: Nothing particularly tripped me up as a user who knows Realm and JavaScript fairly well. However, for beginner users, it's a long sentence with a lot of jargon. I think it would read better if we split it up. For example:

"To apply an update to a collection of objects, iterate through the collection (e.g., using `<Array>.forEach`). In the iterator callback, update each collection individually:"

In this suggestion, performing an iterator and updating each object that's being iterated upon are two separate steps (that are directly correlated to the code block below).

File: source/sdk/node/examples/read-and-write-data.txt (hunk 8 above)
Comment (mongomoe) [reply source/sdk/node/examples/read-and-write-data.txt]: Sounds fine to me 👍

File: source/sdk/node/examples/read-and-write-data.txt (hunk 7 above)
Comment (nlarew) [reply source/sdk/node/examples/read-and-write-data.txt:208]: Yeah the autobuilder has been down so the staging build hasn't updated recently, should hopefully be fixed now. +1 on your suggestion thanks!

File: source/sdk/node/examples/authenticate-users.txt (hunk 10 above)
Comment (nlarew) [reply source/sdk/node/examples/authenticate-users.txt:23]: We mainly use it to refer to anonymous users, but I agree that "temporary" is probably more clear/accessible

File: source/sdk/node/examples/react-to-changes.txt (hunk 21 above)
Comment (nlarew) [reply source/sdk/node/examples/react-to-changes.txt:56]: Yeah you're right some of this should go in the Fundamentals section. This page was a straight port so I can spend some time tomorrow to make it more example-y.

File: source/sdk/node/examples/react-to-changes.txt (hunk 25 above)
Comment (mongomoe) [reply source/sdk/node/examples/react-to-changes.txt:136]: @nlarew can you address this? This is one of the last items before LGTM on this page

Comment (mongomoe) [reply source/sdk/node/examples/react-to-changes.txt:136]: actually looks like its the last item*

Comment (nlarew) [reply source/sdk/node/examples/react-to-changes.txt]: It's updated to `Realm.Object.removeListener()`

**Review by mongomoe — APPROVED** [review]
LGTM on my end. Perhaps send this to #realm-js on slack as well

**Comment by kneth** [comment]
LGTM
